third_party/nvidia/lib/Dialect/NVWS/Transforms/InsertTmemAref.cpp

```diff
@@ -565,11 +571,66 @@ insertTmemArefImpl(TmemAccessDag::Node *node,
   return node;
 }
 
-LogicalResult insertTmemAref(TmemAccessDag &accessDag) {
+bool haveEnoughTmem(MMAv5OpInterface mmaOp, int numTmemBlocks) {
+  auto tmemDesc = mmaOp.getAccumulator().getType();
+  auto blockM = tmemDesc.getShape()[0];
+  auto blockN = tmemDesc.getShape()[1];
+  constexpr int numTMEMColumns = 512;
+  constexpr int numTMEMRows = 128;
+  if (numTmemBlocks + (blockM * blockN * 2) > numTMEMRows * numTMEMColumns) {
+    return false;
+  }
+  if (isa<TCGen5MMAScaledOp>(mmaOp) && blockN == 256) {
+    return false;
+  }
+  return true;
+};
+
+bool hasProducerConsumerPartitioning(TmemAccessDag &accessDag) {
+  // TMEM partitioning follows producer-consumer pattern if it has this
+  // structure:
+  //
+  //      |alloc
+  //      |-- ops
+  //    loop (A) (tt.ws)
+  //      |----  ops (A)
+  //      |----  ops (B)
+  //      |----  possibly ops (A)
+  //      |----  yield (A)
```

Comment:
I don't understand this diagram (what's `loop (A)`, what `possibly ops` could be etc). Can you use a concrete example to describe your idea, for example why the lit test `nested_loop_no_double_buffer` does not fit this pattern.

Reply:
updated comment, and logic. I realized just loop at number of group changes is insufficient, the logic now verifies that one group produces and other consumes.